Debug operator

RuiAAPeres opened the pull request:
Closes #2833

iv-mexx commented on ReactiveCocoa/Swift/SignalProducer.swift:
```diff
@@ -1267,3 +1267,12 @@ private final class DeallocationToken {
+	/// Logs all events that the receiver sends.
+	/// By default, it will print to the standard output.
+	@warn_unused_result(message="Did you forget to call `start` on the producer?")
+	public func debug() -> SignalProducer<Value, Error> {
```
Here, the parameters `identifier` and `logger` are missing.

iv-mexx commented on ReactiveCocoa/Swift/SignalProducer.swift:
```diff
@@ -1267,3 +1267,12 @@ private final class DeallocationToken {
 	}
 }
+
+extension SignalProducerType where Value: Equatable {
```
I think it should not be necessary to constraint `Value` to conform to `Equatable`, or is there?

JaviSoto commented:
Super excited to have this kind of operator in RAC :)

I would recommend a name more along the lines of `logEvents()`, `debug()` is kind of an overloaded word.
This is the extension method I wrote in the Fabric app. It is tightly coupled with the logging framework we use (XCGLogger), and I love the way you abstracted that in your implementation, but I think it would be nice to allow for more customization: https://gist.github.com/JaviSoto/cff1da9049ebdd3ae92e98c5d45b7f15

mdiep commented on ReactiveCocoa/Swift/Signal.swift:
```diff
@@ -1277,3 +1277,35 @@ extension SignalType where Error == NoError {
+///
+/// This is useful, when you want to do more than
+/// just printing to the standard output.
+public protocol EventLogger {
```
I'm not sure I see utility of `EventLogger` when you could just as easily write a `Event<Value, Error> -> ()` with `on(event:)`.

mdiep commented:
> I would recommend a name more along the lines of `logEvents()`, `debug()` is kind of an overloaded word.

👍 to that.

rcgary commented on Documentation/DebuggingTechniques.md:
````diff
@@ -14,6 +14,7 @@ In both cases errors are related to incorrect assumptions about type. Such issue
 Below is an example of type-error scenario:
 
 ```
````
I think you need to delete this line :)

RuiAAPeres commented:
> I would recommend a name more along the lines of logEvents(), debug() is kind of an overloaded word.

👍

@JaviSoto I like your approach with the enum + sets! ✨

> but I think it would be nice to allow for more customization

So I will add the following: 

1. `fileName: String = #file`
2. `functionName: String = #function`
3. `lineNumber: Int = #line`

Do you reckon we should go with the Set as well?

@mdiep the protocol opens room for further development, a `func` in this case would work as well, but I feel the protocol is more appropriate in this case.

RuiAAPeres commented:
Couple of things regarding 4038989: 

1. Moved the operator (utility types + global function) to a different file. This file could be used for other debugging operators.
2. Added @JaviSoto suggestion with the event enum, plus `#filename`, `#functionaName`, `#lineNumber`.

JaviSoto commented:
The reason why I think the eventTypes filter parameter ends up being necessary is that some signals are very noisy, and sometimes you don't care about values, just start / disposal for example. And if you're adding this operator in more than one place in the app, it becomes even more noisy (and for this reason as well, being able to provide some sort of name or identifier is useful)

mdiep commented on ReactiveCocoa/Swift/EventLogger.swift:
```diff
@@ -0,0 +1,82 @@
+///
+/// This is useful, when you want to do more than
+/// just printing to the standard output.
+public protocol EventLogger {
```
Since we don't have a clear use case in mind, I think we should remove `EventLogger` in the initial implementation. It adds complexity and it's much harder to remove API than to add it. (Removing is a breaking change; adding is a non-breaking change in this case.)

There's nothing that prevents anyone from implementation the behavior that this would allow.

RuiAAPeres replied:
I am ok with going with the function, but I am not completely sold. I find it more idiomatic, to with a protocol approach on this.

mdiep commented on ReactiveCocoa/Swift/EventLogger.swift:
```diff
@@ -0,0 +1,82 @@
+}
+
+public enum LoggingEvent: String {
+	case Started, Next, Completed, Failed, Terminated, Disposed, Interrupted
```
Would it make sense to make this an inner `enum` on `Signal` and `SignalProducer`? Then `Started` could be removed as a value for `Signal`.

RuiAAPeres replied:
We are repeating a lot of stuff with that approach. 👎

mdiep commented on ReactiveCocoa/Swift/EventLogger.swift:
```diff
@@ -0,0 +1,82 @@
+		let disposed: OptionalClosure = createEventLog(.Disposed, events: events, logEvent: logEvent)
+		let next: (Self.Value -> Void)? = createEventLog(.Next, events: events, logEvent: logEvent)
+
+		return self.on(failed: failed, completed: completed, interrupted: interrupted, terminated: terminated, disposed: disposed, next: next)
```
It might be clearer to write this like:

```swift
return on(
    failed: createEventLog(.Failed, events: events, logEvent: logEvent),
    completed: createEventLog(.Completed, events: events, logEvent: logEvent),
    …
)
```

I think that would remove the need for the type annotations on the local variables.

mdiep commented on ReactiveCocoa/Swift/EventLogger.swift:
```diff
@@ -0,0 +1,82 @@
+			logger.logEvent("[\(identifier)] \(event)", fileName: fileName, functionName: functionName, lineNumber: lineNumber)
+		}
+		
+		let failed: (Self.Error -> Void)? = createEventLog(.Failed, events: events, logEvent: logEvent)
```
You could probably move `createEventLog` inside the methods so that they can close over `events` and `logEvent`:

```swift
func log<T>(event: LoggingEvent) -> (T -> Void) {
    return events.contains(event) ? { logEvent("\(event.rawValue) \($0)") } : nil    
}
return on(
    failed: log(.Failed),
    …
)
```

RuiAAPeres replied:
Doing that like so, will work well for `Next` and `Error. But it will do the following for all the others:

```swift
Completed ()
Interrupted ()
```

mdiep commented on ReactiveCocoa/Swift/EventLogger.swift:
```diff
@@ -0,0 +1,66 @@
+		
+		func log<T>(event: LoggingEvent) -> (T -> Void)? {
+			return events.contains(event) ? { logEvent("\(event.rawValue) \($0)") } : nil
+		}
```
I think you could add a `Void`-specific `log` function here and below like you had originally to avoid the `()` in the output that you mentioned.

RuiAAPeres replied:
It will need a difference name, I don't think func overload works in nested functions.

mdiep replied:
I think a different name would be fine.

mdiep commented on ReactiveCocoa/Swift/EventLogger.swift:
```diff
@@ -0,0 +1,66 @@
+}
+
+public typealias EventLogger = (event: String, fileName: String, functionName: String, lineNumber: Int) -> Void
+typealias OptionalClosure = (Void -> Void)?
```
This `typealias` isn't needed anymore. 🎉

mdiep commented on ReactiveCocoa/Swift/EventLogger.swift:
```diff
@@ -0,0 +1,66 @@
+		               interrupted: log(.Interrupted),
+		               terminated: log(.Terminated),
+		               disposed: log(.Disposed),
+		               next: log(.Next))
```
It'd be a better match to the RAC style to either put these all on one line or to arrange them like so:

```swift
return self.on(
    failed: log(.Failed),
    completed: log(.Completed),
    interrupted: log(.Interrupted),
    terminated: log(.Terminated),
    disposed: log(.Disposed),
    next: log(.Next)
)
```

mdiep commented on ReactiveCocoa/Swift/EventLogger.swift:
```diff
@@ -0,0 +1,66 @@
+	
+	public static let allEvents: Set<LoggingEvent> = {
+		return [.Started, .Next, .Completed, .Failed, .Terminated, .Disposed, .Interrupted]
+	}()
```
Gonna restart the discussion about separate `LoggingEvent` `enum`s for `Signal` and `SignalProducer` since the previous one was collapsed.

You mentioned repeated code. The cases would be repeated, and `SignalProducer.logEvents` wouldn't be able to `lift, but I don't think it'd be much more code.

I'm concerned that people will expected `Started` to work on `Signal`s when it doesn't. Separate types would let use enforce correctness with the types.

What do you think? (I'm not trying to force you; just trying to have a discussion. :relaxed:)

RuiAAPeres replied:
> You mentioned repeated code.

You will have 2 enums, where the only difference is one case, plus you won't be able to lift. 😕

mdiep replied:
I think that'd be a win. `lift` isn't really any simpler than calling `on` directly.

mdiep commented on ReactiveCocoa/Swift/EventLogger.swift:
```diff
@@ -0,0 +1,66 @@
+	print("\(event) fileName: \(fileName), functionaName: \(functionName), lineNumber: \(lineNumber)")
+}
+
+public typealias EventLogger = (event: String, fileName: String, functionName: String, lineNumber: Int) -> Void
```
Would it make sense to pass a `LoggingEvent` and a generic `Value?` here instead of a `String`?

Then the default `printLog` could combine them like you did in the inner `log` functions, but custom log functions would be a bit more flexible.

RuiAAPeres replied:
How would you define that? Generics in typeesalias are only available in Swift 3. https://github.com/apple/swift-evolution/blob/master/proposals/0048-generic-typealias.md

mdiep replied:
Interesting. I hadn't considered that. Should we drop the `typealias` then? It could be reintroduced with Swift 3 support.

RuiAAPeres replied:
The code without the typealias is unreadable. I understand the goal is to make this as flexible as possible, but I feel that from a code maintenance perspective it's becoming worst.

RuiAAPeres replied:
> Would it make sense to pass a LoggingEvent and a generic Value? here instead of a String?

the event is fine, the problem is with generic value, because we are using the `printLog` as a default value in the `logEvents` func.

RuiAAPeres replied:
I think we should see if this is necessity, before making the code even more complex.

mdiep commented on ReactiveCocoa/Swift/EventLogger.swift:
```diff
@@ -0,0 +1,88 @@
+			interrupted: log(.Interrupted),
+			terminated: log(.Terminated),
+			disposed: log(.Disposed),
+			next: logValue(.Next))
```
Indentation is off a level here.

mdiep commented on ReactiveCocoa/Swift/EventLogger.swift:
```diff
@@ -0,0 +1,88 @@
+			interrupted: log(.Interrupted),
+			terminated: log(.Terminated),
+			disposed: log(.Disposed),
+			next: logValue(.Next))
```
Here too.

RuiAAPeres replied:
It's rendering correctly on my side. 😕 
<img width="311" alt="screen shot 2016-05-02 at 12 21 49" src="https://cloud.githubusercontent.com/assets/949286/14953312/7b8e118c-1060-11e6-9c41-e5f8087b6bf0.png">

mdiep replied:
Oops. My mistake. Not sure what I was thinking. 🙊

mdiep commented:
Just needs a couple indentation fixes now. Otherwise it looks good to me. 👍 ✨

JaviSoto commented on ReactiveCocoa/Swift/EventLogger.swift:
```diff
@@ -0,0 +1,88 @@
+	print("[\(identifier)] \(event) fileName: \(fileName), functionaName: \(functionName), lineNumber: \(lineNumber)")
+}
+
+public typealias EventLogger = (identifier: String, event: String, fileName: String, functionName: String, lineNumber: Int) -> Void
```
It seems to me that this is as much of a type (that we must maintain backwards compatibility with) than an actual protocol with a method, and I think a protocol is a lot easier to work with.

RuiAAPeres replied:
The function approach gives you a bit more flexibility. I am fine with both approaches

JaviSoto commented on ReactiveCocoa/Swift/EventLogger.swift:
```diff
@@ -0,0 +1,88 @@
+	}()
+}
+
+private func printLog(identifier: String, event: String, fileName: String, functionName: String, lineNumber: Int) -> Void {
```
Maybe call this something along the lines of "default event log"?

JaviSoto commented on ReactiveCocoa/Swift/EventLogger.swift:
```diff
@@ -0,0 +1,88 @@
+	/// Logs all events that the receiver sends.
+	/// By default, it will print to the standard output.
+	@warn_unused_result(message="Did you forget to call `observe` on the signal?")
+	public func logEvents(identifier: String = "", events: Set<SignalLoggingEvent> = SignalLoggingEvent.allEvents, fileName: String = #file, functionName: String = #function, lineNumber: Int = #line, logger: EventLogger = printLog) -> Signal<Value, Error> {
```
What do you think about making identifier an explicit parameter name? That way callers have to say:

logEvents(identifier: "foo")

And not just

logEvents("foo")

JaviSoto commented on ReactiveCocoa/Swift/EventLogger.swift:
```diff
@@ -0,0 +1,88 @@
+			logger(identifier: identifier, event: event, fileName: fileName, functionName: functionName, lineNumber: lineNumber)
+		}
+		
+		func log(event: SignalLoggingEvent) -> (Void -> Void)? {
```
Neat!!

JaviSoto commented:
Just a few small nits, but I love this :) thanks a lot!!!

mdiep commented:
Looks good to me. 👍
